clients/subscriber.go

```diff
@@ -0,0 +1,171 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clients
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/conduitio-labs/conduit-connector-gcp-pubsub/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/gammazero/deque"
+	"github.com/google/uuid"
+)
+
+// Subscriber represents a struct with a GCP Pub/Sub client,
+// queues for messages, channels and a function to cancel the context.
+type Subscriber struct {
+	*pubSub
+
+	lock       sync.Mutex
+	msgDeque   *deque.Deque[*pubsub.Message]
+	ackDeque   *deque.Deque[*pubsub.Message]
+	errorCh    chan error
+	canceledCh chan struct{}
+	ctxCancel  context.CancelFunc
+}
+
+// NewSubscriber initializes a new subscriber client and starts receiving a messages to the queue.
+func NewSubscriber(ctx context.Context, cfg config.Source) (*Subscriber, error) {
+	ps, err := newClient(ctx, cfg.General)
+	if err != nil {
+		return nil, fmt.Errorf("new pubsub client: %w", err)
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
+
+	sub := &Subscriber{
+		pubSub:     ps,
+		msgDeque:   deque.New[*pubsub.Message](),
+		ackDeque:   deque.New[*pubsub.Message](),
+		errorCh:    make(chan error),
+		canceledCh: make(chan struct{}),
+		ctxCancel:  cancel,
+	}
+
+	go func() {
+		s := sub.pubSub.client.Subscription(cfg.SubscriptionID)
+
+		// set MaxExtension less than 0, to not extend the ack deadline for each message
+		s.ReceiveSettings.MaxExtension = -1
+
+		if err = s.Receive(cctx,
+			func(_ context.Context, m *pubsub.Message) {
+				sub.lock.Lock()
+				defer sub.lock.Unlock()
+
+				sub.msgDeque.PushBack(m)
+			},
+		); err != nil {
```

Comment:
It might be a matter of preference, but at least to me the following reads better:
```go
err := s.Receieve(...)
if err != nil {
}
```

Reply:
Deal.